src/api/index.ts

```diff
@@ -0,0 +1,191 @@
+import InvalidApiRequestError from "../errors/ApiCallError/InvalidApiRequestError";
+import ApiCallFailedError from "../errors/ApiCallError/ApiCallFailedError";
+import ApiResponse, {
+  OperationsApiResponse,
+  OperationApiResponse
+} from "./ApiResponse";
+
+// TODO: make this configurable, probably by environment variable
+// It will be useful for several things:
+// a) working offline
+// b) testing API-dependent code paths
+// c) different prod/test/staging environments
+const API_BASE = "https://raha-5395e.appspot.com/api/";
+
+/* ==============================
+ * Definitions of how to address
+ * and use API endpoints
+ * ==============================
+ */
+
+/**
+ * Canonical name of an endpoint you can query.
+ */
+export const enum ApiEndpoint {
+  TRUST_MEMBER = "TRUST_MEMBER",
+  GET_OPERATIONS = "GET_OPERATIONS"
+}
+
+/**
+ * Definition for the arguments you need to call a particular API endpoint
+ */
+interface ApiCallDef<E extends ApiEndpoint, Params, Body> {
+  endpoint: E;
+  params: Params;
+  body: Body;
+}
+type TrustMemberApiCall = ApiCallDef<
+  ApiEndpoint.TRUST_MEMBER,
+  { uid: Uid },
+  void
+>;
+type GetOperationsApiCall = ApiCallDef<ApiEndpoint.GET_OPERATIONS, void, void>;
+/**
+ * All API calls you can make, and the arguments you need to call them.
+ */
+export type ApiCall = TrustMemberApiCall | GetOperationsApiCall;
+
+/**
+ * Definition of how to use an API endpoint, i.e. what you have to provide to
+ * call it, and what it will return to you.
+ */
+interface ApiEndpointDef<Call extends ApiCall, Resp extends ApiResponse> {
+  call: Call;
+  response: Resp;
+}
+
+export type TrustMemberApiEndpoint = ApiEndpointDef<
+  TrustMemberApiCall,
+  OperationApiResponse
+>;
+
+export type GetOperationsApiEndpoint = ApiEndpointDef<
+  GetOperationsApiCall,
+  OperationsApiResponse
+>;
+type ApiDefinition = TrustMemberApiEndpoint | GetOperationsApiEndpoint;
+
+/* =================================
+ * Resolving API endpoint locations
+ * =================================
+ */
+enum HttpVerb {
+  GET = "GET",
+  POST = "POST",
+  PUT = "PUT",
+  DELETE = "DELETE",
+  PATCH = "PATCH"
+}
+
+/**
+ * The location of an API endpoint. Uri may contain wildcards that must be
+ * resolved, and only represents a path without a domain to send the request to.
+ */
+interface ApiEndpointSpec {
+  uri: string;
+  method: HttpVerb;
+}
+
+/**
+ * Mapping from API endpoints to their corresponding unresolved locations
+ */
+const apiEndpointSpecs: { [key in ApiEndpoint]: ApiEndpointSpec } = {
+  [ApiEndpoint.TRUST_MEMBER]: {
+    uri: "members/:uid/trust",
+    method: HttpVerb.POST
+  },
+  [ApiEndpoint.GET_OPERATIONS]: {
+    uri: "operations",
+    method: HttpVerb.GET
+  }
+};
+
+/**
+ * The location of an API endpoint, with a full url to query and parameters
+ * resolved.
+ */
+interface ResolvedApiEndpointSpec {
+  url: string;
+  method: HttpVerb;
+}
+
+/**
+ * Determines the URL and HTTP method for an API call.
```

Comment:
Could you add notes on how this resolves URLs? I.e. -- expects named params of the format `:<paramname>`.